cluster-autoscaler/cloudprovider/ovhcloud/ovh_cloud_provider.go

```diff
@@ -0,0 +1,256 @@
+	// GPULabel is the label added to nodes with GPU resource.
+	GPULabel = "node.kubernetes.ovhcloud.com/gpu"
+
+	// NodePoolLabel is the label added to nodes grouped by node group.
+	// Should be soon prepend with `node.kubernetes.ovhcloud.com/`
+	NodePoolLabel = "nodepool"
+
+	// MachineAvailableState defines the state for available flavors for node resources.
+	MachineAvailableState = "available"
+
+	// GPUMachineCategory defines the default instance category for GPU resources.
+	GPUMachineCategory = "t"
+)
+
+// OVHCloudProvider implements CloudProvider interface.
+type OVHCloudProvider struct {
+	manager *OvhCloudManager
+
+	autoscalingOptions config.AutoscalingOptions
+	discoveryOptions   cloudprovider.NodeGroupDiscoveryOptions
+	resourceLimiter    *cloudprovider.ResourceLimiter
+}
+
+// BuildOVHcloud builds the OVHcloud provider.
+func BuildOVHcloud(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	// Open cloud provider folder
+	var configFile io.ReadCloser
+	if opts.CloudConfig != "" {
+		var err error
+
+		configFile, err = os.Open(opts.CloudConfig)
+		if err != nil {
+			klog.Fatalf("Failed to open cloud provider configuration %s: %v", opts.CloudConfig)
+		}
+
+		defer configFile.Close()
+	}
+
+	// Create a new manager given the cloud config previously loaded
+	manager, err := NewManager(configFile)
+	if err != nil {
+		klog.Fatalf("Failed to create OVHcloud manager: %v", err)
+	}
+
+	provider := &OVHCloudProvider{
+		manager: manager,
+
+		autoscalingOptions: opts,
+		discoveryOptions:   do,
+		resourceLimiter:    rl,
+	}
+
+	return provider
+}
+
+// Name returns name of the cloud provider.
+func (provider *OVHCloudProvider) Name() string {
+	return cloudprovider.OVHcloudProviderName
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (provider *OVHCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	groups := make([]cloudprovider.NodeGroup, 0)
+
+	// Cast API node pools into CA node groups
+	// Do not add in array node pool which does not support autoscaling
+	for _, pool := range provider.manager.NodePools {
+		if !pool.Autoscale {
+			continue
+		}
+
+		ng := NodeGroup{
+			NodePool:    pool,
+			Manager:     provider.manager,
+			CurrentSize: -1,
+		}
+
+		groups = append(groups, &ng)
+	}
+
+	return groups
+}
+
+// NodeGroupForNode returns the node group for the given node, nil if the node
+// should not be processed by cluster autoscaler, or non-nil error if such
+// occurred. Must be implemented.
+func (provider *OVHCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	// Fetch node pool label on nodes
+	labels := node.GetLabels()
+	label, exists := labels[NodePoolLabel]
+	if !exists {
+		return &NodeGroup{}, nil
+	}
+
+	// Find in node groups stored in cache which one is linked to the node
+	for _, pool := range provider.manager.NodePools {
+		if pool.Name == label {
```

Comment:
Reading the function comment, shouldn't we add a condition `if !pool.Autoscaled { return nil }` ?